app/src/main/java/io/heckel/ntfy/msg/DownloadIconWorker.kt

```diff
@@ -0,0 +1,193 @@
+                    }
+                }
+                Log.d(TAG, "Icon download: successful response, proceeding with download")
+                save(icon.copy(
+                    size = bytesCopied,
+                    contentUri = uri.toString()
+                ))
+            }
+        } catch (e: Exception) {
+            failed(e)
+
+            // Toast in a Worker: https://stackoverflow.com/a/56428145/1440785
+            val handler = Handler(Looper.getMainLooper())
+            handler.postDelayed({
+                Toast
+                    .makeText(context, context.getString(R.string.detail_item_icon_download_failed, e.message), Toast.LENGTH_LONG)
+                    .show()
+            }, 200)
+        }
+    }
+
+    private fun updateIconFromResponse(response: Response): Icon {
+        val size = if (response.headers["Content-Length"]?.toLongOrNull() != null) {
+            Log.d(TAG, "We got the long! icon here")
+            response.headers["Content-Length"]?.toLong()
+        } else {
+            icon.size // May be null!
+        }
+        val mimeType = if (response.headers["Content-Type"] != null) {
+            response.headers["Content-Type"]
+        } else {
+            val ext = MimeTypeMap.getFileExtensionFromUrl(icon.url)
+            if (ext != null) {
+                val typeFromExt = MimeTypeMap.getSingleton().getMimeTypeFromExtension(ext)
+                typeFromExt ?: icon.type // May be null!
+            } else {
+                icon.type // May be null!
+            }
+        }
+        Log.d(TAG, "New icon size: $size, type: $mimeType")
+        return icon.copy(
+            size = size,
+            type = mimeType
+        )
+    }
+
+    private fun failed(e: Exception) {
+        Log.w(TAG, "Icon download failed", e)
+        maybeDeleteFile()
+    }
+
+    private fun maybeDeleteFile() {
+        val uriCopy = uri
+        if (uriCopy != null) {
+            Log.d(TAG, "Deleting leftover icon $uriCopy")
+            val resolver = applicationContext.contentResolver
+            resolver.delete(uriCopy, null, null)
+        }
+    }
+
+    private fun save(newIcon: Icon) {
+        Log.d(TAG, "Updating icon: $newIcon")
+        icon = newIcon
+        notification = notification.copy(icon = newIcon)
+        notifier.update(subscription, notification)
+        repository.updateNotification(notification)
+    }
+
+    private fun shouldAbortDownload(): Boolean {
+        val maxAutoDownloadSize = MAX_ICON_DOWNLOAD_SIZE
+        val size = icon.size ?: return false // Don't abort if size unknown
+        return size > maxAutoDownloadSize
+    }
+
+    private fun createUri(notification: Notification): Uri {
+        val iconDir = File(context.cacheDir, ICON_CACHE_DIR)
+        if (!iconDir.exists() && !iconDir.mkdirs()) {
+            throw Exception("Cannot create cache directory for icons: $iconDir")
+        }
+        val file = ensureSafeNewFile(iconDir, notification.id)
```

Comment:
You are storing one icon per notification, so there is no re-use at all. We should talk about a caching concept.

Reply:
As mentioned on Discord, your idea of a table `url, contentUri, created` sounds good to me.